File: apps/cms/src/collections/Runs.ts
import type { CollectionConfig } from "payload";

export const Runs: CollectionConfig = {
  slug: "runs",
  admin: {
    useAsTitle: "id",
    defaultColumns: [
      "job_state",
      "id",
      "source",
      // keep DB status but primary state is job_state above
      "status",
      "maxItems",
      "startedAt",
      "completedAt",
    ],
    description: "Individual scraping job executions",
  },
  access: {
    read: () => true,
    create: () => true,
    update: () => true,
    delete: () => true,
  },
  fields: [
    // Computed Job State (from Source.status)
    {
      name: "job_state",
      label: "Job State",
      type: "ui",
      admin: {
        components: {
          Cell: "@/components/RunStateCell",
        },
        description: "active | paused | running (from Source)",
      },
    },
    // Relationships
    {
      name: "source",
      type: "relationship",
      relationTo: "sources",
      required: true,
      label: "Source",
    },

    // Execution Configuration (per-run)
    {
      name: "kind",
      type: "select",
      required: true,
      defaultValue: "manual",
      label: "Execution Type",
      options: [
        { label: "Manual", value: "manual" },
        { label: "Scheduled", value: "scheduled" },
      ],
    },
    {
      name: "maxItems",
      type: "number",
      required: false,
      defaultValue: 0,
      min: 0,
      max: 1000000,
      label: "Max Items for This Run",
      admin: {
        description: "0 means unlimited; otherwise limit per run",
      },
    },
    {
      name: "status",
      type: "select",
      required: true,
      defaultValue: "pending",
      label: "Status",
      options: [
        { label: "Pending", value: "pending" },
        { label: "Running", value: "running" },
        { label: "Paused", value: "paused" },
        { label: "Completed", value: "completed" },
        { label: "Error", value: "error" },
      ],
    },

    // Metrics
    {
      name: "counters",
      type: "json",
      label: "Counters",
      defaultValue: { found: 0, uploaded: 0, errors: 0 },
      admin: {
        description: "Real-time job metrics",
      },
    },

    // Timing
    {
      name: "startedAt",
      type: "date",
      label: "Started At",
      admin: {
        readOnly: true,
      },
    },
    {
      name: "completedAt",
      type: "date",
      label: "Completed At",
      admin: {
        readOnly: true,
      },
    },

    // Error Handling
    {
      name: "errorMessage",
      type: "textarea",
      label: "Error Message",
      admin: {
        condition: (data) => data.status === "error",
      },
    },
  ],
};

File: apps/worker/app/models/runs.py
"""
Runs model - Enhanced with max_items configuration
"""
from sqlalchemy import DateTime, String, Text, JSON, ForeignKey, Integer, Enum
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
from typing import Dict, Any, Optional
import enum

from .base import Base

# Define ENUM types to match Payload CMS
class RunKindEnum(enum.Enum):
    manual = "manual"
    scheduled = "scheduled"

class RunStatusEnum(enum.Enum):
    pending = "pending"
    running = "running"
    paused = "paused"
    completed = "completed"
    error = "error"


class Run(Base):
    __tablename__ = "runs"

    # Primary key - using integer to match Payload
    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True
    )
    
    # Foreign key to source (integer to match)
    source_id: Mapped[int] = mapped_column(
        ForeignKey("sources.id", ondelete="CASCADE"), 
        nullable=False,
        doc="Source that this run belongs to"
    )
    
    # Execution Info (per-run configuration)
    kind: Mapped[RunKindEnum] = mapped_column(
        Enum(RunKindEnum, name="enum_runs_kind", create_type=False),
        default=RunKindEnum.manual,
        nullable=False,
        doc="Execution type: 'manual' or 'scheduled'"
    )
    max_items: Mapped[int] = mapped_column(
        Integer,
        nullable=False,
        default=50,
        doc="Maximum items to scrape for this run"
    )
    status: Mapped[RunStatusEnum] = mapped_column(
        Enum(RunStatusEnum, name="enum_runs_status", create_type=False), 
        default=RunStatusEnum.pending, 
        nullable=False,
        doc="Run status: pending, running, paused, completed, error"
    )
    
    # Counters (matching CMS JSON structure)
    counters: Mapped[Dict[str, Any]] = mapped_column(
        JSON, 
        nullable=True,
        default=lambda: {"found": 0, "uploaded": 0, "errors": 0},
        doc="Run metrics: found, uploaded, errors"
    )
    
    # Timestamps (matching CMS)
    started_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), 
        nullable=True,
        doc="When this run started"
    )
    completed_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), 
        nullable=True,
        doc="When this run completed"
    )
    
    # Error handling (matching CMS)
    error_message: Mapped[str] = mapped_column(
        Text, 
        nullable=True,
        doc="Error message if run failed"
    )
    
    # Payload timestamps
    created_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), 
        nullable=False, 
        server_default=func.now(),
        doc="When this run was created"
    )
    updated_at: Mapped[DateTime] = mapped_column(
        DateTime(timezone=True), 
        nullable=False, 
        server_default=func.now(), 
        onupdate=func.now(),
        doc="When this run was last updated"
    )

    # Relationship to source
    source = relationship("Source", back_populates="runs")

    def __repr__(self) -> str:
        return f"<Run(id={self.id}, source_id={self.source_id}, status='{self.status}', max_items={self.max_items})>"